exporter/stats/prometheus/prometheus.go

```diff
@@ -87,12 +142,182 @@ func (e *Exporter) Export(vd *stats.ViewData) {
 	if len(vd.Rows) == 0 {
 		return
 	}
-	// TODO(jbd,odeke-em): Make sure Export is not blocking
-	// for a long period of time.
-	panic("not implemented")
+
+	e.c.addViewData(vd)
 }
 
 // ServeHTTP serves the Prometheus endpoint.
 func (e *Exporter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	e.handler.ServeHTTP(w, r)
 }
+
+// collector implements prometheus.Collector
+type collector struct {
+	opts Options
+	mu   sync.RWMutex
+
+	// reg helps us keep a reference to our registrar
+	// so that when views are dynamically added, we'll
+	// able to register again without having to expose
+	// an ugly API for the client in which they'll have
+	// to invoke RegisterView for every single view that
+	// they create.
+	reg *prometheus.Registry
+
+	// views are accumulated and atomically
+	// appended to on every Export invocation, from
+	// stats. These views are cleared out when
+	// Collect is invoked and the cycle is repeated.
+	views []*stats.ViewData
+
+	// descs contains the one-time listing of all
+	// descriptions that are retrieved after converting
+	// each view into a prometheus.Metric.
+	//
+	// Note: we use slices here because trying to use channels
+	// with Prometheus.Collector Describe and Collect methods
+	// is quite hairy, moreover for methods that are run once,
+	// yet the count of elements to be collected is unknown
+	// trying to drain our input channel could potential block forever.
+	descs []*prometheus.Desc
+
+	registeredViews map[*stats.View]bool
+
+	// seenMetrics maps from the metric's rawType to the actual Metric.
+	// It is an interface to interface mapping
+	// but the key is the zero value while the value is the instance.
+	seenMetrics map[interface{}]prometheus.Metric
+}
+
+var _ prometheus.Collector = (*collector)(nil)
+
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	for _, desc := range c.descs {
+		ch <- desc
+	}
+}
+
+func (c *collector) lookupMetric(key interface{}) (prometheus.Metric, bool) {
+	c.mu.RLock()
+	value, ok := c.seenMetrics[key]
+	c.mu.RUnlock()
+	return value, ok
+}
+
+func (c *collector) memoizeMetric(key interface{}, value prometheus.Metric) {
+	c.mu.Lock()
+	c.seenMetrics[key] = value
+	c.mu.Unlock()
+}
+
+// Collect fetches the statistics from OpenCensus
+// and delivers them as Prometheus Metrics.
+// Collect is invoked everytime a prometheus.Gatherer is run
+// for example when the HTTP endpoint is invoked by Prometheus.
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	c.mu.Lock()
+	// Get the last views
+	views := c.views
+	// Now clear them out for the next accumulation
+	c.views = c.views[:0]
+	c.mu.Unlock()
+
+	if len(views) == 0 {
+		return
+	}
+
+	seen := make(map[prometheus.Metric]bool)
+	for _, vd := range views {
+		for _, row := range vd.Rows {
+			metric := c.toMetric(vd.View, row)
+			if _, ok := seen[metric]; !ok && metric != nil {
```

Comment:
Won't ok be always true here?

We allocate a new metric each time.  I also couldn't tell why we are skipping some rows here, we should iterate over all of them.

Reply:
No it won't always be true here. For example if we made a DistributionAggregation but then the CountAggregation came later on, ok won't always be true and
also no we don't always allocate a new metric, we have to only create it once
and then reuse it by incrementing, appending a data point to it, inside toMetric.

The reason why we skip over some is because per Collect cycle, if Prometheus sees the same Metric type ie same labels and structural data it'll error out with a duplicate warning. I'll add a comment about this.

Reply:
For example this is what we get:
```
* collected metric opencensus_video_count counter:<value:17 >  was collected before with the same name and label values
```